### PR TITLE
RoundDown function

### PR DESCRIPTION
**Pull request by persiyanov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog.

Category (leave one):

- New Feature

Short description (up to few sentences):

Add `roundDown(x, arr)` function which rounds given number down to one of numbers from given array of boundaries.

Detailed description (optional):

Add `roundDown(x, arr)` function which rounds given number down to one of numbers from given array of boundaries. If `x` less than all numbers in `arr`, it will be rounded to the first element, e.g. `roundDown(-1, [0, 3, 5]) = 0`.

### CODE REVIEW

File: dbms/src/Functions/FunctionsRound.h [hunk 1]
```diff
@@ -556,6 +562,163 @@ class FunctionRounding : public IFunction
 };
 
 
+class FunctionRoundDown : public IFunction {
```
Comment (ilivans) [inline dbms/src/Functions/FunctionsRound.h]: https://clickhouse.yandex/docs/en/development/style/

> 3. Opening and closing curly brackets must be on a separate line

In other places too

File: dbms/src/Functions/FunctionsRound.h [hunk 2]
```diff
@@ -556,6 +562,163 @@ class FunctionRounding : public IFunction
...
+    }
+
+    bool isVariadic() const override { return false; }
+    size_t getNumberOfArguments() const override { return 2; }
+    ColumnNumbers getArgumentsThatAreAlwaysConstant() const override { return {1}; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override {
+        const DataTypePtr & type_x = arguments[0];
+
+        if (!isNumber(type_x))
```
Comment (ilivans) [inline dbms/src/Functions/FunctionsRound.h]: I would recommend to be consistent with one-line if blocks. You should either put all such blocks in brackets or none of them (line 608).

File: dbms/src/Functions/FunctionsRound.h [hunk 3]
```diff
@@ -556,6 +562,163 @@ class FunctionRounding : public IFunction
...
+class FunctionRoundDown : public IFunction {
+public:
+    static constexpr auto name = "roundDown";
+    static FunctionPtr create(const Context & context) { return std::make_shared<FunctionRoundDown>(context); }
+    FunctionRoundDown(const Context & context) : context(context) {}
+
+public:
+    String getName() const override
+    {
+        return name;
```
Comment (ilivans) [inline dbms/src/Functions/FunctionsRound.h]: Could be one-lined as the methods below

File: dbms/src/Functions/FunctionsRound.cpp [hunk 4]
```diff
@@ -10,6 +10,7 @@ void registerFunctionsRound(FunctionFactory & factory)
     factory.registerFunction<FunctionFloor>("floor", FunctionFactory::CaseInsensitive);
     factory.registerFunction<FunctionCeil>("ceil", FunctionFactory::CaseInsensitive);
     factory.registerFunction<FunctionTrunc>("trunc", FunctionFactory::CaseInsensitive);
+    factory.registerFunction<FunctionRoundDown>();
```
Comment (ilivans) [inline dbms/src/Functions/FunctionsRound.cpp]: Why is it registered differently from other functions?

**Review by ilivans — COMMENTED** [review]
- Isn't there any way to make your function match FunctionRounding template (maybe some insignificant changes to FunctionRounding could be enough)?
- Some comments would be nice to see. 
- The function should be added to common comment section (line 36)

File: dbms/src/Functions/FunctionsRound.cpp (hunk 4 above)
Comment (alexey-milovidov) [reply dbms/src/Functions/FunctionsRound.cpp]: This is Ok.
It means default registration - with case sensitive name that function provides itself.

File: dbms/src/Functions/FunctionsRound.h [hunk 5]
```diff
@@ -556,6 +562,163 @@ class FunctionRounding : public IFunction
...
+        {
+            throw Exception{"Second argument of function " + getName() + " must be constant array.", ErrorCodes::ILLEGAL_COLUMN};
+        }
+
+        auto in_column = block.getByPosition(arguments[0]).column;
+        const auto & in_type = block.getByPosition(arguments[0]).type;
+
+        if (in_column->isColumnConst())
+        {
+            executeConst(block, arguments, result, input_rows_count);
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsRound.h]: We can avoid separate branch for constant argument by utilizing
```
    /** If the function have non-zero number of arguments,
      *  and if all arguments are constant, that we could automatically provide default implementation:
      *  arguments are converted to ordinary columns with single value, then function is executed as usual,
      *  and then the result is converted to constant column.
      */
    virtual bool useDefaultImplementationForConstants() const { return false; }

    /** Some arguments could remain constant during this implementation.
      */
    virtual ColumnNumbers getArgumentsThatAreAlwaysConstant() const { return {}; }
```

File: dbms/src/Functions/FunctionsRound.h [hunk 6]
```diff
@@ -556,6 +562,163 @@ class FunctionRounding : public IFunction
...
+        return true;
+    }
+
+    template <typename T>
+    void executeImplNumToNum(const PaddedPODArray<T> & src, PaddedPODArray<T> & dst, const Array & boundaries) {
+        PaddedPODArray<T> bvalues(boundaries.size());
+        for (size_t i = 0; i < boundaries.size(); ++i) {
+            bvalues[i] = boundaries[i].get<T>();
+        }
+
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsRound.h]: TODO: sort boundaries.

File: dbms/src/Functions/FunctionsRound.h [hunk 7]
```diff
@@ -556,6 +562,163 @@ class FunctionRounding : public IFunction
...
+        {
+            in_column = castColumn(block.getByPosition(arguments[0]), return_type, context);
+        }
+
+        const auto in = in_column.get();
+        auto boundaries = array->getValue<Array>();
+        for (size_t i = 0; i < boundaries.size(); ++i) {
+            boundaries[i] = convertFieldToType(boundaries[i], *return_type);
+        }
+
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsRound.h]: Check for empty array.

File: dbms/src/Functions/FunctionsRound.h (hunk 6 above)
Comment (alexey-milovidov) [reply dbms/src/Functions/FunctionsRound.h]: Maybe throw exception if the values are not unique.

File: dbms/src/Functions/FunctionsRound.h [hunk 8]
```diff
@@ -556,6 +562,163 @@ class FunctionRounding : public IFunction
...
+        size_t size = src.size();
+        dst.resize(size);
+        for (size_t i = 0; i < size; ++i)
+        {
+            auto it = std::upper_bound(bvalues.begin(), bvalues.end(), src[i]);
+            if (it == bvalues.end())
+            {
+                dst[i] = bvalues.back();
+            }
+            else if (it == bvalues.begin())
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsRound.h]: At least one of two branches is redundant.

File: dbms/tests/queries/0_stateless/00161_rounding_functions.sql [hunk 9]
```diff
@@ -43,3 +43,6 @@ SELECT 12345.6789 AS x, floor(x, -1), floor(x, -2), floor(x, -3), floor(x, -4),
 
 SELECT roundToExp2(100), roundToExp2(64), roundToExp2(3), roundToExp2(0), roundToExp2(-1);
 SELECT roundToExp2(0.9), roundToExp2(0), roundToExp2(-0.5), roundToExp2(-0.6), roundToExp2(-0.2);
+
+
+SELECT toInt32(number) as x, roundDown(x, [0, 1, 2, 3, 4, 5]) FROM system.numbers LIMIT 10;
```
Comment (alexey-milovidov) [inline dbms/tests/queries/0_stateless/00161_rounding_functions.sql]: Ok.

Please add more tests for all the details in spec.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
Almost perfect. Only minor changes remain.

File: dbms/tests/queries/0_stateless/00161_rounding_functions.sql (hunk 9 above)
Comment (alexey-milovidov) [reply dbms/tests/queries/0_stateless/00161_rounding_functions.sql]: Especially for cases that require type conversions.

**Comment by alexey-milovidov** [comment]
Some minor details remain. I will merge it myself.

**Comment by alexey-milovidov** [comment]
You've edited the test reference file manually and tabs are accidentially replaced by spaces. That's why your test has failed. I will fix it.

**Comment by alexey-milovidov** [comment]
Performance evaluation remains...

**Comment by alexey-milovidov** [comment]
Escpecially performance comparison against `roundAge` and `roundDuration`.

**Comment by alexey-milovidov** [comment]
```
mtlog-perftest03j.yandex.ru :) SELECT count() FROM system.numbers WHERE NOT ignore(roundAge(rand() % 100))

SELECT count()
FROM system.numbers 
WHERE NOT ignore(roundAge(rand() % 100))

↙ Progress: 2.41 billion rows, 19.26 GB (109.80 million rows/s., 878.44 MB/s.) Cancelling query.
Ok.
Query was cancelled.

0 rows in set. Elapsed: 22.023 sec. Processed 2.41 billion rows, 19.26 GB (109.30 million rows/s., 874.42 MB/s.) 

mtlog-perftest03j.yandex.ru :) SELECT count() FROM system.numbers WHERE NOT ignore(roundDown(rand() % 100, [0, 1, 18, 25, 35, 45, 55]))

SELECT count()
FROM system.numbers 
WHERE NOT ignore(roundDown(rand() % 100, [0, 1, 18, 25, 35, 45, 55]))

↓ Progress: 4.85 billion rows, 38.82 GB (85.51 million rows/s., 684.05 MB/s.) Cancelling query.
Ok.
Query was cancelled.

0 rows in set. Elapsed: 56.855 sec. Processed 4.85 billion rows, 38.82 GB (85.35 million rows/s., 682.83 MB/s.)
```

About 20% slower - that's Ok.

**Comment by alexey-milovidov** [comment]
```
mtlog-perftest03j.yandex.ru :) SELECT count() FROM system.numbers WHERE NOT ignore(roundDuration(rand() % 65536))

SELECT count()
FROM system.numbers 
WHERE NOT ignore(roundDuration(rand() % 65536))

↙ Progress: 6.22 billion rows, 49.76 GB (121.83 million rows/s., 974.66 MB/s.) Cancelling query.
Ok.
Query was cancelled.

0 rows in set. Elapsed: 51.152 sec. Processed 6.22 billion rows, 49.76 GB (121.59 million rows/s., 972.74 MB/s.) 

mtlog-perftest03j.yandex.ru :) SELECT count() FROM system.numbers WHERE NOT ignore(roundDown(rand() % 65536, [0, 1, 10, 30, 60, 120, 180, 240, 300, 600, 1200, 1800, 3600, 7200, 18000, 36000]))

SELECT count()
FROM system.numbers 
WHERE NOT ignore(roundDown(rand() % 65536, [0, 1, 10, 30, 60, 120, 180, 240, 300, 600, 1200, 1800, 3600, 7200, 18000, 36000]))

→ Progress: 2.91 billion rows, 23.31 GB (79.74 million rows/s., 637.96 MB/s.) Cancelling query.
Ok.
Query was cancelled.

0 rows in set. Elapsed: 36.640 sec. Processed 2.91 billion rows, 23.31 GB (79.52 million rows/s., 636.19 MB/s.)
```

It's still slower than roundDuration despite the fact that binary search is used instead of linear chain of branches.

**Comment by alexey-milovidov** [comment]
For extra large arrays the performance is down due to array preparation on each block. That's expected (we may use PreparedFunction to make it better).

```
SELECT count() FROM system.numbers_mt WHERE NOT ignore(roundDown(rand() % 65536, (SELECT groupArray(number) FROM numbers(65536))))
```

**Comment by alexey-milovidov** [comment]
Note that there are many optimization possibilities:
- SIMD optimization of comparison;
- better layout for binary search;
- no binary search for small arrays.

**Pull request merged**